src/lifecycleEvents.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { AnyJson, Dictionary } from '@salesforce/ts-types';
+import * as Debug from 'debug';
+
+type callback = (data: AnyJson) => Promise<void>;
+
+/**
+ * An asynchronous event listener and emitter that follows the singleton pattern.
```

Comment:
Put some more context here with `@examples` like some of the other classes in force. Something like:

An asynchronous event listener and emitter that follows the singleton pattern. The singleton pattern allows lifecycle events to be emitted from deep within a library and still be consumed by any other library or tool. It allows other developers to react to certain situations or events in your library without them having to manually call the method themselves.

An example might be transforming metadata before it is deployed to an environment. As long as an event was emitted from the deploy library and you were listening on that event in the same process, you could transform the metadata before regardless of where in the code that metadata was initiated. 

@example
// Listen for an event in a plugin hook
Lifecycle.getInstance().on('deploy-metadata', transformMetadata)

// Deep in the deploy code, fire the event for all libraries and plugins to hear.
Lifecycle.getInstance().emit('deploy-metadata', metadataToBeDeployed);


Or something like that....